### PR TITLE
feat: Add spec.ordinals.start.

### PR DESCRIPTION
**Pull request by vimystic**
* Added spec.ordinal.start field to FullNodeSpec
* Ran make generate & make manifests to automodify required files.
* Updated BuildPods to accept startingOrdinal.
* Made supporting changes for peer , configmap & service.

Completes https://github.com/strangelove-ventures/cosmos-operator/issues/445

Example usage within our patchfiles for sentries & fullnodes. 

Screenshot 2024-12-04 at 12 11 00 PM

Now tested in project colo-primary-test and cluster vcluster-fullnode-testnet.
This is a test with starting ordinal 9.
Please ignroe the Pending pod & pvc.
That is because there are only 2 functional workers in the test cluster.
    
Screenshot 2024-11-20 at 4 17 46 AM
Screenshot 2024-11-20 at 4 17 58 AM

Also tested instanceOverrides with a PVC example and a disabling pod. 
Screenshot 2024-11-27 at 2 14 48 AM

### CODE REVIEW

**Comment by vimystic** [comment]
Performed a test locally with minikube with startingOrdinal Value 5. 
Screenshot 2024-11-18 at 4 23 37 AM

**Comment by vimystic** [comment]
Now tested in project colo-primary-test and cluster vcluster-fullnode-testnet. 
This is a test with starting ordinal 9. 
Please ignroe the Pending pod. That is because there are only 2 functional workers in the test cluster.
Screenshot 2024-11-20 at 4 17 46 AM
Screenshot 2024-11-20 at 4 17 58 AM

File: api/v1/cosmosfullnode_types.go
Comment (nourspace) [inline api/v1/cosmosfullnode_types.go:39]: From StatefulSet

> start is the number representing the first replica's index. It may be used to number replicas from an alternate index (eg: 1-indexed) over the default 0-indexed names, or to orchestrate progressive movement of replicas from one StatefulSet to another. If set, replica indices will be in the range: [.spec. ordinals. start, .spec. ordinals. start + .spec. replicas). If unset, defaults to 0. Replica indices will be in the range: [0, .spec. replicas).

File: api/v1/cosmosfullnode_types.go
Comment (nourspace) [inline api/v1/cosmosfullnode_types.go:47]: From StatefulSet

> ordinals controls the numbering of replica indices in a StatefulSet. The default ordinals behavior assigns a "0" index to the first replica and increments the index by one for each additional replica requested.

Comment (nourspace) [reply api/v1/cosmosfullnode_types.go:47]: Source: https://www.manifests.io/kubernetes/1.31/io.k8s.api.apps.v1.StatefulSetOrdinals?linked=StatefulSet.spec.ordinals

Comment (vimystic) [reply api/v1/cosmosfullnode_types.go:47]: Updated such that we have it as `spec.ordinals.start` instead of  `spec.ordinal.start`.

File: api/v1/cosmosfullnode_types.go
Comment (vimystic) [reply api/v1/cosmosfullnode_types.go:39]: Yes, the othere references in the code , configmap, pvc builders ensure that the range is  (spec.ordinals.start, spec.ordinals.start + spec.replicas)

**Review by nourspace — CHANGES_REQUESTED** [review]
Update descriptions to match that of Statefulset

File: config/crd/bases/cosmos.strange.love_cosmosfullnodes.yaml
Comment (nourspace) [inline config/crd/bases/cosmos.strange.love_cosmosfullnodes.yaml:558]: This should be `ordinals`

PS. There are two descriptions.
- ordinals
- start

Please copy these from above comments.

**Comment by pharr117** [comment]
@vimystic and I did a code review walkthrough and this is looking good to me, some comments:

1. We fixed the `ordinal` vs `ordinals` issue by modifying the struct tags
2. We discussed the possibility of integer overflow for the int32 ordinal value and found it not too in need of attention unless someone else concerns as well
3. Update the for loops for code consistency across all usages of the ordinal loop

I will approve once Nour's comments are fully addressed as well

File: config/crd/bases/cosmos.strange.love_cosmosfullnodes.yaml
Comment (vimystic) [reply config/crd/bases/cosmos.strange.love_cosmosfullnodes.yaml]: Thanks for this catch , the json tag needed to be corrected

**Comment by vimystic** [comment]
Working on fixing lint/test errors.

**Review by pharr117 — APPROVED** [review]

File: api/v1/cosmosfullnode_types.go
Comment (vimystic) [reply api/v1/cosmosfullnode_types.go]: Updated descriptions appropriately in sturct and autogenerated file.

**Review by nourspace — COMMENTED** [review]
add happy path test for pod creation

**Comment by vimystic** [comment]
I cannot merge until this is gone.
Screenshot 2024-12-12 at 7 20 54 PM

**Comment by vimystic** [comment]
Screenshot 2024-12-12 at 7 21 39 PM
I cannot merge until the above is removed.

**Review by nourspace — APPROVED** [review]
Good work!

**Pull request merged**
